controllers/Handle.php

```diff
@@ -57,15 +58,26 @@ public function __construct()
     public function callback(string $provider): RedirectResponse
     {
         if (!in_array($provider, $this->enabledProviders)) {
-            abort(404);
+            return $this->redirectToSigninPage("The {$provider} SSO provider is not enabled");
         }
 
         // @TODO: Login or register the user / provide an event for plugins to handle
         // user registration themselves. Would like plugin to be able to handle frontend
         // or backend or even both. If event is used follow naming conventions from in progress
         // issues
 
-        $ssoUser = Socialite::driver($provider)->user();
+        if (!Request::input('code')) {
+            $message = 'Error: no access token was returned by provider (' . $provider . ')';
+            if ($error = Request::input('error')) {
+                $message = $provider . ' error: ' . $error;
+                if ($errorDescription = Request::input('error_description')) {
+                    $message .= ' (' . $errorDescription . ')';
+                }
+            }
+            return $this->redirectToSigninPage($message);
+        }
```

Comment:
Where do the `error` and `error_description` values come from? Is that a standard feature of Socialite or is it specific to a particular provider?

Reply:
I think this is in oAuth protocol, all providers seem to return those. I also found this:

https://www.oauth.com/oauth2-servers/access-tokens/access-token-response/